src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
-                        prune(dropPeer, topic)
+
+                // drop underscored peers from mesh
+                peers.filter { score.score(it) < 0 }
+                    .forEach { prune(it, topic) }
+
+                if (peers.size < coreParams.DLow) {
+                    // need more mesh peers
+                    (getTopicPeers(topic) - peers)
+                        .filter { score.score(it) >= 0 && !isDirect(it) && !isBackOff(it, topic) }
+                        .shuffled(random)
+                        .take(coreParams.D - peers.size)
+                        .forEach { graft(it, topic) }
+                } else if (peers.size > coreParams.DHigh) {
+                    // too many mesh peers
+                    val sortedPeers = peers
+                        .shuffled(random)
+                        .sortedBy { score.score(it) }
+                        .reversed()
+
+                    val bestDPeers = sortedPeers.take(coreParams.DScore)
+                    val restPeers = sortedPeers.drop(coreParams.DScore).shuffled(random)
+                    val outboundCount = bestDPeers.count { it.isOutbound() }
+                    val outPeers = restPeers
+                        .filter { it.isOutbound() }
+                        .take(max(0, coreParams.DOut - outboundCount))
+
+                    val toDropPeers = (outPeers + bestDPeers + restPeers).drop(coreParams.DScore)
+                    toDropPeers.forEach { prune(it, topic) }
+                }
+
+                // keep outbound peers > DOut
+                val outboundCount = peers.count { it.isOutbound() }
+                (getTopicPeers(topic) - peers)
+                    .filter { it.isOutbound() && score.score(it) >= 0 && !isDirect(it) && !isBackOff(it, topic) }
+                    .shuffled(random)
+                    .take(max(0, coreParams.DOut - outboundCount))
+                    .forEach { graft(it, topic) }
+
+                // opportunistic grafting
+                if (heartbeatsCount % scoreParams.opportunisticGraftTicks == 0 && peers.size > 1) {
+                    val scoreMedian = peers.map { score.score(it) }.median()
+                    if (scoreMedian < scoreParams.opportunisticGraftThreshold) {
+                        (getTopicPeers(topic) - peers)
+                            .filter { score.score(it) > scoreMedian && !isDirect(it) && !isBackOff(it, topic) }
+                            .take(params.opportunisticGraftPeers)
+                            .forEach { graft(it, topic) }
                     }
                 }
-                submitGossip(topic, peers)
+
+                emitGossip(topic, peers)
             }
             fanout.entries.forEach { (topic, peers) ->
-                peers.removeIf { it in getTopicPeers(topic) }
-                val needMore = D - peers.size
+                peers.removeIf {
+                    it !in getTopicPeers(topic) || score.score(it) < scoreParams.publishThreshold
```

Comment:
The spec suggests that `publishThreshold` is only used when flood publishing. Is fanout a type of flood publishing? 
> publishThreshold: when a peer's score drops below this threshold, self published messages are not propagated towards this peer when (flood) publishing.

Reply:
Hmm, good question! 
That condition was taken from the Go implementation: https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L1440-L1442
But I also didn't find any mentions in the spec. 
@vyzo may be you can comment?

Reply:
Yeah, the publish threshold applies everywhere there is an actual publication.